File: Source/Extensions/Tessa.Extensions.Default.Client/UI/WorkflowEngine/KrRoutesInWorkflowEngineUIExtension.cs
using System.Threading.Tasks;
using System.Windows;

using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.UI.Cards;

using static Tessa.Extensions.Default.Shared.Workflow.KrProcess.KrConstants;

namespace Tessa.Extensions.Default.Client.UI.WorkflowEngine
{
    /// <summary>
    /// Настраивает отображение карточки в которой используются маршруты в Workflow Engine.
    /// </summary>
    public sealed class KrRoutesInWorkflowEngineUIExtension
        : CardUIExtension
    {
        #region Fields

        private readonly IKrTypesCache typesCache;

        #endregion

        #region Constructor

        /// <summary>
        /// Инициализирует новый экземпляр класса <see cref="KrRoutesInWorkflowEngineUIExtension"/>.
        /// </summary>
        /// <param name="typesCache">Кэш по типам карточек и документов, содержащих информацию по типовому решению.</param>
        public KrRoutesInWorkflowEngineUIExtension(
            IKrTypesCache typesCache)
        {
            this.typesCache = typesCache;
        }

        #endregion

        #region Base overrides

        /// <inheritdoc/>
        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var model = context.Model;
            var card = model.Card;
            var cardTypeID = card.TypeID;

            if (model.Flags.Has(CardModelFlags.EditTemplate)
                || KrProcessSharedHelper.DesignTimeCard(cardTypeID)
                || (await KrComponentsHelper.GetKrComponentsAsync(
                        card,
                        this.typesCache,
                        context.CancellationToken)).HasNot(KrComponents.Routes)
                || !(await KrProcessSharedHelper.TryGetKrTypeAsync(
                        this.typesCache,
                        card,
                        cardTypeID,
                        validationResult: context.ValidationResult,
                        validationObject: this,
                        cancellationToken: context.CancellationToken)).UseRoutesInWorkflowEngine
                || !model.Forms.TryGet(Ui.KrApprovalProcessFormAlias, out var routesForm)
                || !model.Blocks.TryGet(Ui.KrApprovalStagesBlockAlias, out var stageBlock))
            {
                return;
            }

            if (stageBlock.BlockVisibility != Visibility.Collapsed)
            {
                stageBlock.BlockVisibility = Visibility.Collapsed;
                routesForm.RearrangeSelf();
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/workflowEngine/krRoutesInWorkflowEngineUIExtension.ts
import {
  CardModelFlags,
  CardUIExtension,
  IBlockViewModel,
  ICardUIExtensionContext,
  IFormWithBlocksViewModel
} from 'tessa/ui/cards';
import { hasFlag, hasNotFlag, Visibility } from 'tessa/platform';
import { getKrComponentsByCard, KrComponents, KrTypesCache } from 'tessa/workflow';
import { designTimeCard, tryGetKrType } from '../../workflow/krProcess/krUIHelper';

export default class KrRoutesInWorkflowEngineUIExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext) {
    const model = context.model;
    const card = model.card;
    const cardTypeId = card.typeId;
    const routesForm: IFormWithBlocksViewModel | undefined = model.forms.find(
      x => x.name === 'ApprovalProcess'
    );
    const stageBlock: IBlockViewModel | undefined = model.blocks.get('ApprovalStagesBlock');
    const hasNotRoutes = hasNotFlag(
      getKrComponentsByCard(card, KrTypesCache.instance),
      KrComponents.Routes
    );
    const krType = tryGetKrType(KrTypesCache.instance, card, cardTypeId, context.validationResult);
    if (
      hasFlag(model.flags, CardModelFlags.EditTemplate) ||
      designTimeCard(cardTypeId) ||
      hasNotRoutes ||
      (krType && !krType.useRoutesInWorkflowEngine) ||
      !routesForm ||
      !stageBlock
    ) {
      return;
    }
    if (stageBlock.blockVisibility !== Visibility.Collapsed) {
      stageBlock.blockVisibility = Visibility.Collapsed;
    }
  }
}
